File: SundeeFundee/Sources/SundeeFundeeKit/UI/ViewModels/AnalyticsViewModel.swift
import Foundation

// MARK: - AnalyticsViewModel

/// View model that bridges the persistence layer to analytics chart views.
///
/// Fetches Workout, OneRepMaxRecord, and CyclePhaseInfo from the data client,
/// transforms them through ChartDataAggregator, and publishes chart-ready
/// data points. All analytics are available because Sundee Fundee is free.
@available(iOS 18.0, macOS 15.0, watchOS 11.0, *)
@MainActor
public class AnalyticsViewModel: ObservableObject {

    // MARK: - Published State

    /// Currently selected time range for all charts.
    @Published public var selectedTimeRange: TimeRange = .lastSixMonths {
        didSet {
            if selectedTimeRange != oldValue {
                reaggregate()
            }
        }
    }

    /// Currently selected exercise for the strength progression chart.
    /// When nil, all exercises are shown.
    @Published public var selectedExercise: String? {
        didSet {
            filterStrengthData()
        }
    }

    /// Strength progression data points for the selected exercise (or all).
    @Published public var strengthData: [StrengthDataPoint] = []

    /// Training volume data points aggregated by week.
    @Published public var volumeData: [VolumeDataPoint] = []

    /// Workout frequency data points aggregated by week.
    @Published public var frequencyData: [FrequencyDataPoint] = []

    /// Cycle-correlated performance data points.
    @Published public var cycleData: [CyclePerformancePoint] = []

    /// Unique exercise names available for the exercise picker.
    @Published public var availableExercises: [String] = []

    /// High-level progress summary shown before deeper charts.
    @Published public var progressSnapshot: ProgressSnapshot?
    @Published public var cycleInsights: [CycleAwareProgressInsight] = []

    /// Whether data is currently being fetched.
    @Published public var isLoading: Bool = false

    /// Error message to display if data fetching fails.
    @Published public var errorMessage: String?

    /// Whether there is no data to display (all sources empty).
    public var hasNoData: Bool {
        !isLoading && errorMessage == nil
            && allWorkouts.isEmpty && allORMRecords.isEmpty && allCyclePhases.isEmpty
    }

    // MARK: - Dependencies

    private let dataClient: DataClientProtocol

    // MARK: - Cached Raw Data

    /// All one-rep-max records fetched from persistence (unfiltered).
    private var allORMRecords: [OneRepMaxRecord] = []

    /// All workouts fetched from persistence (unfiltered).
    private var allWorkouts: [Workout] = []

    /// All cycle phase info records fetched from persistence (unfiltered).
    private var allCyclePhases: [CyclePhaseInfo] = []

    /// All challenges fetched from persistence (unfiltered).
    private var allChallenges: [Challenge] = []

    // MARK: - Initialization

    public init(
        dataClient: DataClientProtocol = DataClientFactory.shared.client
    ) {
        self.dataClient = dataClient
    }

    // MARK: - Public Methods

    /// Fetches all data from persistence and populates chart data via ChartDataAggregator.
    public func loadAnalytics() async {
        isLoading = true
        errorMessage = nil

        do {
            // Fetch all record types in parallel
            async let ormFetch: [OneRepMaxRecord] = dataClient.fetchAll(
                recordType: "OneRepMaxRecord"
            )
            async let workoutFetch: [Workout] = dataClient.fetchAll(
                recordType: "Workout"
            )
            async let phaseFetch: [CyclePhaseInfo] = dataClient.fetchAll(
                recordType: "CyclePhaseInfo"
            )

            let (ormRecords, workouts, phases) = try await (ormFetch, workoutFetch, phaseFetch)
            let challenges: [Challenge] = (try? await dataClient.fetchAll(recordType: "Challenge")) ?? []

            // Cache raw data
            allORMRecords = ormRecords
            allWorkouts = workouts
            allCyclePhases = phases
            allChallenges = challenges

            // Populate available exercises
            availableExercises = ChartDataAggregator.exercises(from: ormRecords)

            // Reaggregate with current time range
            reaggregate()
        } catch {
            errorMessage = error.localizedDescription
            // Clear chart data on error — graceful degradation
            strengthData = []
            volumeData = []
            frequencyData = []
            cycleData = []
            availableExercises = []
            progressSnapshot = nil
            cycleInsights = []
        }

        isLoading = false
    }

    /// Filters the strength progression chart to a single exercise.
    ///
    /// Pass nil to show all exercises.
    public func selectExercise(_ name: String?) {
        selectedExercise = name
    }

    // MARK: - Private Methods

    /// Re-runs all aggregators with the current time range and cached data.
    private func reaggregate() {
        let range = selectedTimeRange

        // Strength progression
        let allStrength = ChartDataAggregator.strengthProgression(
            from: allORMRecords,
            timeRange: range
        )

        // Apply exercise filter
        if let exercise = selectedExercise {
            strengthData = allStrength.filter { $0.exerciseName == exercise }
        } else {
            strengthData = allStrength
        }

        // Training volume
        volumeData = ChartDataAggregator.trainingVolume(
            from: allWorkouts,
            timeRange: range
        )

        // Workout frequency
        frequencyData = ChartDataAggregator.workoutFrequency(
            from: allWorkouts,
            timeRange: range
        )

        // Cycle correlation
        cycleData = ChartDataAggregator.cycleCorrelation(
            from: allWorkouts,
            phases: allCyclePhases,
            timeRange: range
        )

        progressSnapshot = ProgressSnapshotService.build(
            maxRecords: allORMRecords,
            workouts: allWorkouts,
            challenges: allChallenges,
            cycleData: cycleData
        )
        cycleInsights = CycleAwareProgressInsightService.build(
            cycleData: cycleData,
            workouts: allWorkouts,
            maxRecords: allORMRecords
        )
    }

    /// Filters strength data to the selected exercise without re-fetching.
    private func filterStrengthData() {
        let allStrength = ChartDataAggregator.strengthProgression(
            from: allORMRecords,
            timeRange: selectedTimeRange
        )

        if let exercise = selectedExercise {
            strengthData = allStrength.filter { $0.exerciseName == exercise }
        } else {
            strengthData = allStrength
        }
    }
}

File: android/app/src/main/java/com/sundeefundee/app/viewmodel/AnalyticsViewModel.kt
package com.sundeefundee.app.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sundeefundee.core.data.factory.DataClientFactory
import com.sundeefundee.core.domain.analytics.ChartDataAggregator
import com.sundeefundee.core.domain.analytics.TimeRange
import com.sundeefundee.core.model.CyclePhaseInfo
import com.sundeefundee.core.model.OneRepMaxRecord
import com.sundeefundee.core.model.Workout
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class AnalyticsViewModel @Inject constructor(
    private val dataClientFactory: DataClientFactory
) : ViewModel() {

    private val _isLoading = MutableStateFlow(false)
    val isLoading = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage = _errorMessage.asStateFlow()

    private val _selectedRange = MutableStateFlow(TimeRange.THREE_MONTHS)
    val selectedRange = _selectedRange.asStateFlow()

    private val _strengthData = MutableStateFlow<List<ChartDataAggregator.StrengthDataPoint>>(emptyList())
    val strengthData = _strengthData.asStateFlow()

    private val _volumeData = MutableStateFlow<List<ChartDataAggregator.VolumeDataPoint>>(emptyList())
    val volumeData = _volumeData.asStateFlow()

    private val _frequencyData = MutableStateFlow<List<ChartDataAggregator.FrequencyDataPoint>>(emptyList())
    val frequencyData = _frequencyData.asStateFlow()

    private val _selectedExercise = MutableStateFlow<String?>(null)
    val selectedExercise = _selectedExercise.asStateFlow()

    private val _exerciseNames = MutableStateFlow<List<String>>(emptyList())
    val exerciseNames = _exerciseNames.asStateFlow()

    fun loadData() {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val workouts: List<Workout> = dataClientFactory.client.fetchAll("Workout")
                val maxes: List<OneRepMaxRecord> = dataClientFactory.client.fetchAll("OneRepMaxRecord")

                _exerciseNames.value = workouts.flatMap { it.exercises.map { ex -> ex.name } }.distinct().sorted()

                _strengthData.value = ChartDataAggregator.strengthProgression(maxes, _selectedRange.value)
                _volumeData.value = ChartDataAggregator.trainingVolume(workouts, _selectedRange.value)
                _frequencyData.value = ChartDataAggregator.workoutFrequency(workouts, _selectedRange.value)
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
            _isLoading.value = false
        }
    }

    fun selectRange(range: TimeRange) {
        _selectedRange.value = range
        loadData()
    }

    fun selectExercise(name: String?) {
        _selectedExercise.value = name
    }
}
